shadow-4.0.3/man/grpck.8
.\"$Id: grpck.8 6 2005-03-20 15:34:28Z bubulle $
.\" Copyright 1992 - 1993, Julianne Frances Haugh
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of Julianne F. Haugh nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY JULIE HAUGH AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL JULIE HAUGH OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.TH GRPCK 8
.SH NAME
grpck \- verify integrity of group files
.SH SYNOPSIS
\fBgrpck\fR [\fB-r\fR] [\fIgroup\fR \fIshadow\fR]
.SH DESCRIPTION
\fBgrpck\fR verifies the integrity of the system authentication information.
All entries in the \fI/etc/group\fR and \fI/etc/gshadow\fR are checked to
see that the entry has the proper format and valid data in each field.
The user is prompted to delete entries that are improperly formatted or
which have other incorrectable errors.
.P
Checks are made to verify that each entry has
.sp
.in +.5i
- the correct number of fields
.br
- a unique group name
.br
- a valid list of members and administrators
.in -.5i
.sp
.P
The checks for correct number of fields and unique group name are fatal.
If the entry has the wrong number of fields, the user will be prompted to
delete the entire line.
If the user does not answer affirmatively, all further checks are bypassed.
An entry with a duplicated group name is prompted for deletion, but the
remaining checks will still be made.
All other errors are warnings and the user is encouraged to run the
\fBgroupmod\fR command to correct the error.
.P
The commands which operate on the \fI/etc/group\fR file are not able to
alter corrupted or duplicated entries.
\fBgrpck\fR should be used in those circumstances to remove the offending
entry.
.SH OPTIONS
By default, \fBgrpck\fR operates on the files \fI/etc/group\fR and
\fI/etc/gshadow\fR.
The user may select alternate files with the \fIgroup\fR and \fIshadow\fR
parameters.
Additionally, the user may execute the command in read-only mode by
specifying the \fB-r\fR flag.
This causes all questions regarding changes to be answered \fBno\fR
without user intervention.
\fBgrpck\fR can also sort entries in \fI/etc/group\fR and \fI/etc/gshadow\fR
by GID. To run it in sort mode pass it \fB-s\fR flag. No checks are
performed then, it just sorts.
.SH FILES
/etc/group \- group account information
.br
/etc/gshadow \- encrypted passwords and group administrator information
.br
/etc/passwd \- user information
.SH SEE ALSO
.BR group (5),
.BR passwd (5),
.BR shadow (5),
.BR groupmod (8)
.SH DIAGNOSTICS
The \fBgrpck\fR command exits with the following values:
.IP 0 5
Success
.IP 1 5
Syntax Error
.IP 2 5
One or more bad group entries
.IP 3 5
Cannot open group files
.IP 4 5
Cannot lock group files
.IP 5 5
Cannot update group files
.SH AUTHOR
Julianne Frances Haugh (jockgrrl@ix.netcom.com)


noflushd/NEWS
* noflushd 2.7.5 (09. May 2005)

  - Fix handling of raid and similar meta devices on 2.6 kernels.

* noflushd 2.7.4 (08. Aug 2004)

  - Fix lousy spindown times on 2.6 kernels.
  - Fix SCSI support on 2.6 kernels.
  - Fix pid detection in generic startup script.
  - Document quirks with hotplugged drives.

* noflushd 2.7.3 (09. Apr 2004)

  - Fix race condition that might leave kupdated/pdflushd disabled
    aftern noflushd has terminated.
  - Fix spindown when all disks are auto-detected.

* noflushd 2.7.2 (16. Mar 2004)

  - A change in 2.7.1 erroneously flushed the buffer cache before
    spindown, causing frequent spinups. This change has been reverted.
    
* noflushd 2.7.1 (07. Mar 2004)

  - Fix immediate spinup on 2.6 kernels.
  - Elevate SCSI support from compile-time to command line option.
  - Bug fixes in the disk statistics parser for 2.6 kernels.
  - Ship kernel patch to track down which processes cause spinups.

* noflushd 2.7 (20. Jan 2004)

  - Support Linux kernel 2.6.x.
  - Default to monitorring all disks unless specific disks are given
    on the command line. This change should yield a sane default
    behaviour for a majority of users without requiring explicit
    configuration.

* noflushd 2.6.3 (13. May 2002)

  - Bugfix for arm and other archs with weird /proc/partitions layout.

* noflushd 2.6.2 (20. Apr 2002)

  - Add IDE raid support.
  - Add timeout logging script.

* noflushd 2.6.1 (03. Feb. 2002)

  - Fix BUG when using noflushd with disks on second IDE channel on 2.4
    kernels.

* noflushd 2.6 (22. Jan 2002)

  - Document noflushd's suckage with any journaling fs, including ext3.
  - New feature prevents spindown if machine is in interactive use.
  - Add hack fix some '0 spindown time' uglies.

* noflushd 2.5 (06. Aug 2001)

  - Make sure hotplugged disks get synced properly.
  - Only sync present partitions. (Fixes high CPU usage bug on devfs
    systems.)
  - Generic startup skript works on latest 2.4 kernels.

* noflushd 2.4 (13. Mar 2001)

  - Timeout lists are back to a working state.
  - Support very latest 2.4 kernels.

* noflushd 2.3 (12. Jan 2001)

  - Support latest 2.4 kernels. (Leaves kernels 2.3.99pre8 to 2.4.0-test8
    unsupported.)
  - Fix multi-disk support with long bdflush timeouts.

* noflushd 2.2 (22. Dec 2000)

  - Fixed bug in spindown code. Now actually spins down. Duh!

* noflushd 2.1 (20. Dec 2000)

  - Disks get synced prior to spindown.
  - Improved devfs support.

* noflushd 2.0 (14. Nov 2000)

  - noflushd now comes with its own syncer that takes over the task from
    kupdate when the first read-write mounted disk is spun down. This
    enhances spindown times in a multidisk environment at the expense of
    less smooth and more bursty background flushing when the noflushd 
    syncer is in use.

  - Conflicting entries in 2.2-style /proc/stat (e.g. /dev/sda and /dev/hda)
    won't cause noflushd to give up anymore. Instead, noflushd will treat
    both disks pretty much like one, i.e. they're either both up or spun
    down, no matter if activity only goes to one of them.

  - Compilation environment switched to automake/autoconf.

  - When installing via 'make install', noflushd is no longer added to
    any runlevel. 

  - noflushd moved to SourceForge. The project homepage is located at
    http://noflushd.sourceforge.net, a mailing list is available via
    noflushd-devel@lists.sourceforge.net. Visit the homepage for 
    subscription information.



shadow-4.0.3/src/grpck.c
/*
 * Copyright 1992 - 1994, Julianne Frances Haugh
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of Julianne F. Haugh nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY JULIE HAUGH AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL JULIE HAUGH OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <config.h>

#include "rcsid.h"
RCSID (PKG_VER "$Id: grpck.c 6 2005-03-20 15:34:28Z bubulle $")
#include <stdio.h>
#include <fcntl.h>
#include <grp.h>
#include "prototypes.h"
#include "defines.h"
#include "chkname.h"
#include <pwd.h>
#include "commonio.h"
#include "groupio.h"
extern void __gr_del_entry (const struct commonio_entry *);
extern struct commonio_entry *__gr_get_head (void);

#ifdef SHADOWGRP
#include "sgroupio.h"
extern void __sgr_del_entry (const struct commonio_entry *);
extern struct commonio_entry *__sgr_get_head (void);
#endif

/*
 * Exit codes
 */

#define	E_OKAY		0
#define	E_USAGE		1
#define	E_BAD_ENTRY	2
#define	E_CANT_OPEN	3
#define	E_CANT_LOCK	4
#define	E_CANT_UPDATE	5

/*
 * Local variables
 */

static char *Prog;
static const char *grp_file = GROUP_FILE;

#ifdef	SHADOWGRP
static const char *sgr_file = SGROUP_FILE;
#endif
static int read_only = 0;

/* local function prototypes */
static void usage (void);
static int yes_or_no (void);
static void delete_member (char **, const char *);

/*
 * usage - print syntax message and exit
 */

static void usage (void)
{
#ifdef	SHADOWGRP
	fprintf (stderr, _("Usage: %s [-r] [-s] [group [gshadow]]\n"),
		 Prog);
#else
	fprintf (stderr, _("Usage: %s [-r] [-s] [group]\n"), Prog);
#endif
	exit (E_USAGE);
}

/*
 * yes_or_no - get answer to question from the user
 */

static int yes_or_no (void)
{
	char buf[80];

	/*
	 * In read-only mode all questions are answered "no".
	 */

	if (read_only) {
		puts (_("No"));
		return 0;
	}

	/*
	 * Get a line and see what the first character is.
	 */

	if (fgets (buf, sizeof buf, stdin))
		return buf[0] == 'y' || buf[0] == 'Y';

	return 0;
}

/*
 * delete_member - delete an entry in a list of members
 */

static void delete_member (char **list, const char *member)
{
	int i;

	for (i = 0; list[i]; i++)
		if (list[i] == member)
			break;

	if (list[i])
		for (; list[i]; i++)
			list[i] = list[i + 1];
}

/*
 * grpck - verify group file integrity
 */

int main (int argc, char **argv)
{
	int arg;
	int errors = 0;
	int deleted = 0;
	int i;
	int prune = 0;
	struct commonio_entry *gre, *tgre;
	struct group *grp;
	int sort_mode = 0;

#ifdef	SHADOWGRP
	struct commonio_entry *sge, *tsge;
	struct sgrp *sgr;
	int is_shadow = 0;
#endif

	/*
	 * Get my name so that I can use it to report errors.
	 */

	Prog = Basename (argv[0]);

	setlocale (LC_ALL, "");
	bindtextdomain (PACKAGE, LOCALEDIR);
	textdomain (PACKAGE);

	OPENLOG (Prog);

	/*
	 * Parse the command line arguments
	 */

	while ((arg = getopt (argc, argv, "qprs")) != EOF) {
		switch (arg) {
		case 'q':
			/* quiet - ignored for now */
			break;
		case 'r':
			read_only = 1;
			break;
		case 's':
			sort_mode = 1;
			break;
		case 'p':
			prune = 1;
			break;
		default:
			usage ();
		}
	}

	if (sort_mode && read_only) {
		fprintf (stderr, _("%s: -s and -r are incompatibile\n"),
			 Prog);
		exit (E_USAGE);
	}

	/*
	 * Make certain we have the right number of arguments
	 */

#ifdef	SHADOWGRP
	if (optind != argc && optind + 1 != argc && optind + 2 != argc)
#else
	if (optind != argc && optind + 1 != argc)
#endif
		usage ();

	/*
	 * If there are two left over filenames, use those as the group and
	 * group password filenames.
	 */

	if (optind != argc) {
		grp_file = argv[optind];
		gr_name (grp_file);
	}
#ifdef	SHADOWGRP
	if (optind + 2 == argc) {
		sgr_file = argv[optind + 1];
		sgr_name (sgr_file);
		is_shadow = 1;
	} else if (optind == argc)
		is_shadow = sgr_file_present ();
#endif

	/*
	 * Lock the files if we aren't in "read-only" mode
	 */

	if (!read_only) {
		if (!gr_lock ()) {
			fprintf (stderr, _("%s: cannot lock file %s\n"),
				 Prog, grp_file);
			if (optind == argc)
				SYSLOG ((LOG_WARN, "cannot lock %s",
					 grp_file));
			closelog ();
			exit (E_CANT_LOCK);
		}
#ifdef	SHADOWGRP
		if (is_shadow && !sgr_lock ()) {
			fprintf (stderr, _("%s: cannot lock file %s\n"),
				 Prog, sgr_file);
			if (optind == argc)
				SYSLOG ((LOG_WARN, "cannot lock %s",
					 sgr_file));
			closelog ();
			exit (E_CANT_LOCK);
		}
#endif
	}

	/*
	 * Open the files. Use O_RDONLY if we are in read_only mode,
	 * O_RDWR otherwise.
	 */

	if (!gr_open (read_only ? O_RDONLY : O_RDWR)) {
		fprintf (stderr, _("%s: cannot open file %s\n"), Prog,
			 grp_file);
		if (optind == argc)
			SYSLOG ((LOG_WARN, "cannot open %s", grp_file));
		closelog ();
		exit (E_CANT_OPEN);
	}
#ifdef	SHADOWGRP
	if (is_shadow && !sgr_open (read_only ? O_RDONLY : O_RDWR)) {
		fprintf (stderr, _("%s: cannot open file %s\n"), Prog,
			 sgr_file);
		if (optind == argc)
			SYSLOG ((LOG_WARN, "cannot open %s", sgr_file));
		closelog ();
		exit (E_CANT_OPEN);
	}
#endif

	if (sort_mode) {
		gr_sort ();
#ifdef	SHADOWGRP
		if (is_shadow)
			sgr_sort ();
#endif
		goto write_and_bye;
	}

	/*
	 * Loop through the entire group file.
	 */

	for (gre = __gr_get_head (); gre; gre = gre->next) {
		/*
		 * Skip all NIS entries.
		 */

		if (gre->line[0] == '+' || gre->line[0] == '-')
			continue;

		/*
		 * Start with the entries that are completely corrupt. They
		 * have no (struct group) entry because they couldn't be
		 * parsed properly.
		 */

		if (!gre->eptr) {

			/*
			 * Tell the user this entire line is bogus and ask
			 * them to delete it.
			 */

			printf (_("invalid group file entry\n"));
			printf (_("delete line `%s'? "), gre->line);
			errors++;

			/*
			 * prompt the user to delete the entry or not
			 */
			if (!prune) {
			        if (!yes_or_no ())
				        continue;
			} else {
			        puts("Yes");
				continue;
			}

			/*
			 * All group file deletions wind up here. This code
			 * removes the current entry from the linked list.
			 * When done, it skips back to the top of the loop
			 * to try out the next list element.
			 */

		      delete_gr:
			SYSLOG ((LOG_INFO, "delete group line `%s'",
				 gre->line));
			deleted++;

			__gr_del_entry (gre);
			continue;
		}

		/*
		 * Group structure is good, start using it.
		 */

		grp = gre->eptr;

		/*
		 * Make sure this entry has a unique name.
		 */

		for (tgre = __gr_get_head (); tgre; tgre = tgre->next) {

			const struct group *ent = tgre->eptr;

			/*
			 * Don't check this entry
			 */

			if (tgre == gre)
				continue;

			/*
			 * Don't check invalid entries.
			 */

			if (!ent)
				continue;

			if (strcmp (grp->gr_name, ent->gr_name) != 0)
				continue;

			/*
			 * Tell the user this entry is a duplicate of
			 * another and ask them to delete it.
			 */

			puts (_("duplicate group entry\n"));
			printf (_("delete line `%s'? "), gre->line);
			errors++;

			/*
			 * prompt the user to delete the entry or not
			 */

			if (yes_or_no ())
				goto delete_gr;
		}

		/*
		 * Check for invalid group names.  --marekm
		 */
		if (!check_group_name (grp->gr_name)) {
			errors++;
			printf (_("invalid group name `%s'\n"),
				grp->gr_name);
		}

		/*
		 * Workaround for a NYS libc 5.3.12 bug on RedHat 4.2 -
		 * groups with no members are returned as groups with one
		 * member "", causing grpck to fail.  --marekm
		 */

		if (grp->gr_mem[0] && !grp->gr_mem[1]
		    && *(grp->gr_mem[0]) == '\0')
			grp->gr_mem[0] = (char *) 0;

		/*
		 * Make sure each member exists
		 */

		for (i = 0; grp->gr_mem[i]; i++) {
			if (getpwnam (grp->gr_mem[i]))
				continue;
			/*
			 * Can't find this user. Remove them
			 * from the list.
			 */

			errors++;
			printf (_("group %s: no user %s\n"),
				grp->gr_name, grp->gr_mem[i]);
			printf (_("delete member `%s'? "), grp->gr_mem[i]);

			if (!yes_or_no ())
				continue;

			SYSLOG ((LOG_INFO, "delete member `%s' group `%s'",
				 grp->gr_mem[i], grp->gr_name));
			deleted++;
			delete_member (grp->gr_mem, grp->gr_mem[i]);
			gre->changed = 1;
			__gr_set_changed ();
		}
	}

#ifdef	SHADOWGRP
	if (!is_shadow)
		goto shadow_done;

	/*
	 * Loop through the entire shadow group file.
	 */

	for (sge = __sgr_get_head (); sge; sge = sge->next) {

		/*
		 * Start with the entries that are completely corrupt. They
		 * have no (struct sgrp) entry because they couldn't be
		 * parsed properly.
		 */

		if (!sge->eptr) {

			/*
			 * Tell the user this entire line is bogus and ask
			 * them to delete it.
			 */

			printf (_("invalid shadow group file entry\n"));
			printf (_("delete line `%s'? "), sge->line);
			errors++;

			/*
			 * prompt the user to delete the entry or not
			 */

			if (!yes_or_no ())
				continue;

			/*
			 * All shadow group file deletions wind up here. 
			 * This code removes the current entry from the
			 * linked list. When done, it skips back to the top
			 * of the loop to try out the next list element.
			 */

		      delete_sg:
			SYSLOG ((LOG_INFO, "delete shadow line `%s'",
				 sge->line));
			deleted++;

			__sgr_del_entry (sge);
			continue;
		}

		/*
		 * Shadow group structure is good, start using it.
		 */

		sgr = sge->eptr;

		/*
		 * Make sure this entry has a unique name.
		 */

		for (tsge = __sgr_get_head (); tsge; tsge = tsge->next) {

			const struct sgrp *ent = tsge->eptr;

			/*
			 * Don't check this entry
			 */

			if (tsge == sge)
				continue;

			/*
			 * Don't check invalid entries.
			 */

			if (!ent)
				continue;

			if (strcmp (sgr->sg_name, ent->sg_name) != 0)
				continue;

			/*
			 * Tell the user this entry is a duplicate of
			 * another and ask them to delete it.
			 */

			puts (_("duplicate shadow group entry\n"));
			printf (_("delete line `%s'? "), sge->line);
			errors++;

			/*
			 * prompt the user to delete the entry or not
			 */

			if (yes_or_no ())
				goto delete_sg;
		}

		/*
		 * Make sure this entry exists in the /etc/group file.
		 */

		if (!gr_locate (sgr->sg_name)) {
			puts (_("no matching group file entry\n"));
			printf (_("delete line `%s'? "), sge->line);
			errors++;
			if (yes_or_no ())
				goto delete_sg;
		}

		/*
		 * Make sure each administrator exists
		 */

		for (i = 0; sgr->sg_adm[i]; i++) {
			if (getpwnam (sgr->sg_adm[i]))
				continue;
			/*
			 * Can't find this user. Remove them
			 * from the list.
			 */

			errors++;
			printf (_
				("shadow group %s: no administrative user %s\n"),
				sgr->sg_name, sgr->sg_adm[i]);
			printf (_("delete administrative member `%s'? "),
				sgr->sg_adm[i]);

			if (!yes_or_no ())
				continue;

			SYSLOG ((LOG_INFO,
				 "delete admin `%s' from shadow group `%s'",
				 sgr->sg_adm[i], sgr->sg_name));
			deleted++;
			delete_member (sgr->sg_adm, sgr->sg_adm[i]);
			sge->changed = 1;
			__sgr_set_changed ();
		}

		/*
		 * Make sure each member exists
		 */

		for (i = 0; sgr->sg_mem[i]; i++) {
			if (getpwnam (sgr->sg_mem[i]))
				continue;

			/*
			 * Can't find this user. Remove them from the list.
			 */

			errors++;
			printf (_("shadow group %s: no user %s\n"),
				sgr->sg_name, sgr->sg_mem[i]);
			printf (_("delete member `%s'? "), sgr->sg_mem[i]);

			if (!yes_or_no ())
				continue;

			SYSLOG ((LOG_INFO,
				 "delete member `%s' from shadow group `%s'",
				 sgr->sg_mem[i], sgr->sg_name));
			deleted++;
			delete_member (sgr->sg_mem, sgr->sg_mem[i]);
			sge->changed = 1;
			__sgr_set_changed ();
		}
	}

      shadow_done:
#endif				/* SHADOWGRP */

	/*
	 * All done. If there were no deletions we can just abandon any
	 * changes to the files.
	 */

	if (deleted) {
	      write_and_bye:
		if (!gr_close ()) {
			fprintf (stderr, _("%s: cannot update file %s\n"),
				 Prog, grp_file);
			exit (E_CANT_UPDATE);
		}
#ifdef	SHADOWGRP
		if (is_shadow && !sgr_close ()) {
			fprintf (stderr, _("%s: cannot update file %s\n"),
				 Prog, sgr_file);
			exit (E_CANT_UPDATE);
		}
#endif
	}

	/*
	 * Don't be anti-social - unlock the files when you're done.
	 */

#ifdef	SHADOWGRP
	if (is_shadow)
		sgr_unlock ();
#endif
	(void) gr_unlock ();

	/*
	 * Tell the user what we did and exit.
	 */

	if (errors)
#ifdef NDBM
		printf (deleted ?
			_
			("%s: the files have been updated; run mkpasswd\n")
			: _("%s: no changes\n"), Prog);
#else
		printf (deleted ?
			_("%s: the files have been updated\n") :
			_("%s: no changes\n"), Prog);
#endif

	exit (errors ? E_BAD_ENTRY : E_OKAY);
}
